sparta/test/Memory/Memory_test.cpp

```diff
@@ -793,3 +795,143 @@ void testMemoryObjectSizes() {
+    }
+
+    // Get a DMI deeper into a memory region
+    dmi = membif.getDMI(1024, BLOCK_SIZE);
+    EXPECT_NOTEQUAL(dmi, nullptr);
+    ::memset(buf, 0, BLOCK_SIZE);
+    EXPECT_NOTHROW(dmi->read(1024, BLOCK_SIZE, buf));
+    for (uint32_t i = 0; i < BLOCK_SIZE; ++i) {
+        EXPECT_EQUAL(buf[i], 0xFF);
+    }
+
+    // Get an unaligned DMI.  This will span a block (64 bytes)
+    dmi = membif.getDMI(122, BLOCK_SIZE);
+    EXPECT_EQUAL(dmi, nullptr);
+
+    // Get a just-barely unaligned DMI.  This will span a block (64 bytes)
+    dmi = membif.getDMI(124, 8);
+    EXPECT_EQUAL(dmi, nullptr);
+
+    // This will return a DMI starting at the lowest BLOCK address
+    // away -- [64,128)
+    dmi = membif.getDMI(112, 10);
+    EXPECT_NOTEQUAL(dmi, nullptr);
+    ::memset(buf, 0, BLOCK_SIZE);
+    EXPECT_NOTHROW(dmi->read(112, 10, buf));
+    for (uint32_t i = 0; i < 10; ++i) {
+        EXPECT_EQUAL(buf[i], 0xFF);
+    }
+
+    ::memset(buf, 0, BLOCK_SIZE);
+    // Go outside the window -- start at base address and read beyond the line
+    EXPECT_THROW(dmi->read(64, BLOCK_SIZE + 1, buf));
+    for (uint32_t i = 0; i < BLOCK_SIZE; ++i) {
+        EXPECT_EQUAL(buf[i], 0); // The data should NOT have been read
+    }
+
+    // Read within the line
+    EXPECT_NOTHROW(dmi->read(64, BLOCK_SIZE-10, buf));
+    for (uint32_t i = 0; i < BLOCK_SIZE-10; ++i) {
+        EXPECT_EQUAL(uint32_t(buf[i]), 0xFF); // Data shifted down by 10 elements
+    }
+
+    // Get a DMI deeper into a memory region
+    dmi = membif.getDMI(1024, BLOCK_SIZE);
+    EXPECT_NOTEQUAL(dmi, nullptr);
+    ::memset(buf, 0, BLOCK_SIZE);
+    EXPECT_NOTHROW(dmi->read(1024, BLOCK_SIZE, buf));
+    for (uint32_t i = 0; i < BLOCK_SIZE; ++i) {
+        EXPECT_EQUAL(buf[i], 0xFF);
+    }
+
+    // Get an unaligned DMI.  This will span a block (64 bytes)
+    dmi = membif.getDMI(122, BLOCK_SIZE);
+    EXPECT_EQUAL(dmi, nullptr);
+
+    // Get a just-barely unaligned DMI.  This will span a block (64 bytes)
+    dmi = membif.getDMI(124, 8);
+    EXPECT_EQUAL(dmi, nullptr);
+
+    dmi = membif.getDMI(112, 10);
+    EXPECT_NOTEQUAL(dmi, nullptr);
+    ::memset(buf, 0, BLOCK_SIZE);
+    EXPECT_NOTHROW(dmi->read(112, 10, buf));
+    for (uint32_t i = 0; i < 10; ++i) {
+        EXPECT_EQUAL(buf[i], 0xFF);
+    }
+
+    // Go back to the first access and make sure nothing's changed
+    dmi = membif.getDMI(0, BLOCK_SIZE);
+    EXPECT_NOTEQUAL(dmi, nullptr);
+    EXPECT_EQUAL(dmi, first_dmi);
+    ::memset(buf, 0, BLOCK_SIZE);
+    EXPECT_NOTHROW(dmi->read(0, BLOCK_SIZE, buf));
+    for (uint32_t i = 0; i < BLOCK_SIZE; ++i) {
+        EXPECT_EQUAL(buf[i], i);
+    }
+
+    // Some silly things -- the following lines of code get around the
+    // private getDMI method in DMIBlockingMemoryIF
+    sparta::memory::BlockingMemoryIF * dmi_blocking_if = dmi;
+    EXPECT_THROW(dmi_blocking_if->getDMI(0, BLOCK_SIZE));
+
+    membif.invalidateAllDMI();
```

Comment:
Wondering what's the scenario to invalidate DMI ? Does it mean users have to check the availability of DMI before accessing it?

Reply:
Technically you should always check.   The scenario mostly includes software moving memory around and updating memory maps.  If you originally got a DMI in the core to say ROM @ 0x1000, and software remaps the ROM to 0x3000 and puts something else at 0x1000, the DMI pointer the core has is technically invalid.  It's a DMI pointer to 0x1000, which _still points to ROM_ now at 0x3000.  If you want to access memory at 0x1000, you need to get a new DMI.

Reply:
Gotcha. We're talking about physical memory right? Didn't have much experience in this, but is it possible to change the memory map of devices in runtime?

Reply:
Yep.  Some memory controllers I've worked on had programmable maps.